lib/core/src/Cardano/Wallet/DB/Sqlite.hs

```diff
@@ -505,19 +505,18 @@ newDBLayer trace defaultFieldValues fp = do
                     pure $ Right ()
 
         , readWalletMeta = \(PrimaryKey wid) -> do
-            walDelegation <- delegationFromEntity . fmap entityVal
-                <$> selectFirst [CertWalletId ==. wid] [Desc CertSlot]
-
-            fmap (metadataFromEntity walDelegation . entityVal)
-                <$> selectFirst [WalId ==. wid] []
-
-        , readWalletDelegations = \(PrimaryKey wid) -> ExceptT $ do
-            selectWallet wid >>= \case
-                Nothing -> pure $ Left $ ErrNoSuchWallet wid
-                Just _ -> do
-                    Right . fmap (delegationDiscoveredFromEntity . entityVal)
-                        . take 2
-                        <$> selectList [CertWalletId ==. wid] [Desc CertSlot]
+            selectLatestCheckpoint wid >>= \case
+                Nothing -> pure Nothing
+                Just cp -> do
+                    latestMaybe <- fmap entityVal
+                        <$> selectFirst [CertWalletId ==. wid] [Desc CertSlot]
+                    case latestMaybe of
+                        Just dlg -> do
+                            let (W.SlotId currentEpoch _) = checkpointSlot cp
+                            determineWalletDelegation wid dlg currentEpoch
+                        Nothing ->
+                            retrieveWalletMetadata wid
+                            (delegationFromCerts Nothing Nothing)
```

Comment:
I am not sure to understand the distinction here? Why is there 2 paths for figuring out what the delegation status is?

Reply:
right, moved downstream